### PR TITLE
CRW-352 check if user has accidentally put the server image tag in the server image name field

### PR DESCRIPTION
**Pull request by nickboldt**
CRW-352 check if user has accidentally put the server image tag in the server image name field

Change-Id: I292bf990b72d79f8cc90ea9a3ad385d65ab17bfb
Signed-off-by: nickboldt <nboldt@redhat.com>

### CODE REVIEW

**Comment by rhopp** [comment]
Shouldn't we just unify the parameters and remove `--version` at all? To me, this switch has been bit confusing right from the start.

**Comment by nickboldt** [comment]
I wouldn't remove API (or commandine flags) in a maintenance release. I'll break old docs. I'd rather fix the inconsistency so that users can ignore the --version flag and just use --server-image=image:tag.

It's also possible that some users / jobs / tests are using the default image and ONLY setting `--version=1.2-x`, which we don't want to suddenly break,

File: operator-installer/deploy.sh [hunk 1]
```diff
@@ -95,6 +113,27 @@ export OPENSHIFT_PROJECT=${OPENSHIFT_PROJECT:-${DEFAULT_OPENSHIFT_PROJECT}}
 
 export ENABLE_OPENSHIFT_OAUTH=${ENABLE_OPENSHIFT_OAUTH:-${DEFAULT_ENABLE_OPENSHIFT_OAUTH}}
 
+# CRW 352 check if user has accidentally put the server image tag in the server image name field
+if [[ $SERVER_IMAGE_NAME ]]; then
+  if [[ ! $SERVER_IMAGE_TAG ]]; then
+    SERVER_IMAGE_TAG_MAYBE=${SERVER_IMAGE_NAME#*/*:*}
+    if [[ ${SERVER_IMAGE_NAME} != ${SERVER_IMAGE_TAG_MAYBE} ]]; then
+      SERVER_IMAGE_TAG=${SERVER_IMAGE_TAG_MAYBE}
```
Comment (dmytro-ndp) [inline operator-installer/deploy.sh]: IMHO, `--version` option should have higher priority compare to tag, extracted from image address.

File: operator-installer/deploy.sh [hunk 2]
```diff
@@ -95,6 +113,27 @@ export OPENSHIFT_PROJECT=${OPENSHIFT_PROJECT:-${DEFAULT_OPENSHIFT_PROJECT}}
...
+      SERVER_IMAGE_NAME=${SERVER_IMAGE_NAME%:${SERVER_IMAGE_TAG_MAYBE}}
```
Comment (dmytro-ndp) [inline operator-installer/deploy.sh]: `SERVER_IMAGE_NAME` has already had tag name inside. So, what is the propose of this command?

File: operator-installer/deploy.sh [hunk 3]
```diff
@@ -95,6 +113,27 @@ export OPENSHIFT_PROJECT=${OPENSHIFT_PROJECT:-${DEFAULT_OPENSHIFT_PROJECT}}
...
+      printWarning "Server image tag ${SERVER_IMAGE_TAG} found for server image name ${SERVER_IMAGE_NAME}"
+    fi
+  elif [[ $SERVER_IMAGE_TAG ]]; then
+    SERVER_IMAGE_TAG_MAYBE=${SERVER_IMAGE_NAME#*/*:*}
+    if [[ ${SERVER_IMAGE_NAME} != ${SERVER_IMAGE_TAG_MAYBE} ]]; then 
+      SERVER_IMAGE_NAME=${SERVER_IMAGE_NAME%:${SERVER_IMAGE_TAG_MAYBE}}
+      printError "Server image tag ${SERVER_IMAGE_TAG_MAYBE} found for server image name ${SERVER_IMAGE_NAME}, but also set server image tag = ${SERVER_IMAGE_TAG}."
```
Comment (dmytro-ndp) [inline operator-installer/deploy.sh]: Error message sounds over complicated and misleading, as for me.
Can we print something simpler like "--version option doesn't comply with server image tag ${SERVER_IMAGE_TAG_MAYBE}"?

**Comment by dmytro-ndp** [comment]
@rhopp: without `--version` parameter we should re-configuring all the pre-release testing jobs for 1.2.2/1.2.3 QA cycle (it requires adding tag to option "--server-image=quay.io/crw/server-rhel8").
And we might take into account coming replacement of **deploy.sh** by **chectl**

File: operator-installer/deploy.sh (hunk 3 above)
Comment (nickboldt) [reply operator-installer/deploy.sh]: "compliance" suggests they have to adhere to an API. 

I'm just saying "you set the tag twice. Pick one." :D

File: operator-installer/deploy.sh (hunk 2 above)
Comment (nickboldt) [reply operator-installer/deploy.sh]: To remove the tag from the image name, so that it's not pushed into the custom-resource.yaml.

Comment (dmytro-ndp) [reply operator-installer/deploy.sh]: Got it, thanks.

File: operator-installer/deploy.sh (hunk 3 above)
Comment (dmytro-ndp) [reply operator-installer/deploy.sh]: Okay.

**Comment by nickboldt** [comment]
Merged to 6.19.x and master branches for 1.2 and 2.0

**Pull request merged**